### PR TITLE
Implement selective batching for vllm

### PR DESCRIPTION
**Pull request by gc-fu**
## Description

Implement selective batching for vLLM.

User can use an argument `VLLM_ENABLE_SELECTIVE_BATCHING` to control whether to enable `selective_batching` feature

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/models/llama.py [hunk 1]
```diff
@@ -376,3 +538,175 @@ def native_sdp(query, key, value, attention_mask,
...
+    position_ids: Optional[torch.LongTensor] = None,
+    past_key_values: Optional[List[torch.FloatTensor]] = None,
+    inputs_embeds: Optional[torch.FloatTensor] = None,
+    use_cache: Optional[bool] = None,
+    output_attentions: Optional[bool] = None,
+    output_hidden_states: Optional[bool] = None,
+    return_dict: Optional[bool] = None,
+) -> Union[Tuple, BaseModelOutputWithPast]:
+    if output_attentions is not None:
+        output_attentions = output_attentions
```
Comment (xiangyuT) [inline python/llm/src/bigdl/llm/transformers/models/llama.py:556]: What's the meaning of this line?

File: python/llm/src/bigdl/llm/transformers/models/llama.py [hunk 2]
```diff
@@ -376,3 +538,175 @@ def native_sdp(query, key, value, attention_mask,
...
+            new_mask = self._prepare_decoder_attention_mask(
+                attn_mask, (1, seq_length), inputs_embeds, past_key_value_length
+            )
+            attention_mask[i] = new_mask
+            i += 1
+
+    hidden_states = inputs_embeds
+
+    if self.gradient_checkpointing and self.training:
+        if use_cache:
```
Comment (xiangyuT) [inline python/llm/src/bigdl/llm/transformers/models/llama.py:648]: invalidInputError

File: python/llm/src/bigdl/llm/transformers/models/llama.py [hunk 3]
```diff
@@ -376,3 +538,175 @@ def native_sdp(query, key, value, attention_mask,
...
+                hidden_states,
+                attention_mask,
+                position_ids,
+                None,
+            )
+        else:
+            layer_outputs = decoder_layer(
+                hidden_states,
+                # TODO: decide if we need this attention_mask,
+                # we are not using the attention mask when decoding
```
Comment (xiangyuT) [inline python/llm/src/bigdl/llm/transformers/models/llama.py:682]: delete this

File: python/llm/src/bigdl/llm/transformers/models/llama.py [hunk 4]
```diff
@@ -310,6 +311,167 @@ def llama_attention_forward_4_31(
...
+            current_value_states = torch.cat([past_v,
+                                              value_states[batch: batch + 1, :, :, :]], dim=2)
+
+            updated_past_key_values.append((current_key_states, current_value_states))
+
+            current_key_states = repeat_kv(current_key_states, self.num_key_value_groups)
+            current_value_states = repeat_kv(current_value_states, self.num_key_value_groups)
+
+            current_query_states = query_states[batch: batch + 1, :, :, :]
+            attn_output, aattn_weights = native_sdp(current_query_states,
```
Comment (xiangyuT) [inline python/llm/src/bigdl/llm/transformers/models/llama.py:413]: attn_weights

File: python/llm/src/bigdl/llm/vllm/model_executor/models/bigdl_llama.py [hunk 5]
```diff
@@ -207,8 +243,12 @@ def forward(
         # tmp = torch.xpu.memory_stats()
         # logger.info(f"before: {tmp['allocated_bytes.all.current']}")
 
-        self.update_kv_cache(cur_seq_ids,
-                             kv_cache, num_layers, decoder_kv_size)
+        if enable_vllm_se_batching:
+            self.update_kv_cache_selective_batching(
+                cur_seq_ids, kv_cache, num_layers, decoder_kv_size)
+            self.last_kv_cache = None
```
Comment (xiangyuT) [inline python/llm/src/bigdl/llm/vllm/model_executor/models/bigdl_llama.py:249]: Just do not set last_kv_cache if enable_vllm_se_batching?

**Review by glorysdj — APPROVED** [review]
LGTM

**Review by xiangyuT — APPROVED** [review]
LGTM

**Pull request merged**